Documentation on API types and classes (#62198)
zephyrproject-rtos/zephyr: doc/develop/api/api_conventions.rst

.. _api_conventions.rst:

API Conventions
###############

Zephyr provides a few programming interfaces that users can use for creating their applications.
Zephyr's API space includes the following areas:

- C programming interface (function prototypes, structures, and macros), provided in C header files
- Configuration system (Kconfig)
- Hardware description system (Devicetree)

Other areas, such as the build system (CMake), may also be seen as Zephyr's API. However, any area
that is not explicitly listed herein is not considered Zephyr's API as of today.

API classification
******************
The files that define the Zephyr APIs may contain symbols intended for different usage. The intended
API usage is designated by the API class. Zephyr APIs are classified as:

Private

>     These APIs are intended for use within boundary of a :term:`software component`. Private APIs
>     defined in Zephyr tree are not subject to :ref:`api_lifecycle`. Therefore, can be changed or
>     removed at any time. Changes to the private API may not be documented.

Reviewer: ```suggestion
    These APIs are intended for use within the boundary of a :term:`software component`. Private APIs
    defined in the main Zephyr tree are not subject to :ref:`api_lifecycle`. Therefore, they can be changed or
    removed at any time. Changes to the private APIs may not be documented at all, and are not included in the 
    migration guide.
```

Internal

>     In general, these APIs are intended for use only between certain software components that are
>     located in Zephyr tree. The context where the API is called or implemented is well defined. For
>     example, functions prefixed with `arch_` are intended for use by the Zephyr kernel to invoke
>     architecture specific code. An API is classified as an Internal on a case by case basis.
>     Internal APIs should not be used by out-of-tree code. Internal APIs are not subject
>     to :ref:`api_lifecycle`. Therefore, can be changed or removed at any time. Changes to the
>     internal APIs must be documented in the migration guide.

Reviewer: ```suggestion
    In general, these APIs are intended for use only between certain software components that are
    located in the main Zephyr tree. The context where the API is called or implemented is well defined. For
    example, functions prefixed with `arch_` are intended for use by the Zephyr kernel to invoke
    architecture-specific code. An API is classified as internal on a case by case basis.
    Internal APIs should not be used by out-of-tree code. Internal APIs are not subject
    to :ref:`api_lifecycle`. Therefore, can be changed or removed at any time. However, changes to the
    internal APIs must be documented in the migration guide.
```

Public

>     These APIs intended for use from any :term:`software component`. Public APIs may be used in-tree
>     and out-of-tree. Public APIs are subject to the :ref:`api_lifecycle`. Therefore, changes to an
>     API are introduced and documented according to the rules defined for the API life state.

Reviewer: ```suggestion
    These APIs intended for use from any :term:`software component`. Public APIs may be used in-tree
    and out-of-tree. Public APIs are subject to the :ref:`api_lifecycle`. Therefore, changes to an
    API are introduced and documented according to the rules defined for the API life cycle. This includes 
    documenting any breaking changes in the migration guide.
```

> Note, that only APIs defined in Zephyr tree are subject to :ref:`api_lifecycle`. External projects

Reviewer: ```suggestion
Note, that only APIs defined in the main Zephyr tree are subject to :ref:`api_lifecycle`. External projects
```

used as :ref:`modules` may define their own rules for API lifecycle.

Zephyr is a constantly evolving project and API classification may change over time. A Private or
Internal API may be promoted to Internal or Public API, respectively. Zephyr users are encouraged to
follow :ref:`rfcs` process to recommend changes in API classification.

The following sections provide guidelines on how to identify the class of an API depending on its
type.

Header files contents
=====================
Private
    Functions and data types declared in header files located in
    :zephyr_file:`include/zephyr/private/`. In addition, private symbols are prefixed with ``z_``.
    Due to historical reasons some APIs prefixed with ``z_`` are public.

Internal
    Functions and data types declared :zephyr_file:`include/zephyr/internal`. In addition, Internal
    APIs must use ``@internal`` doxygen command.

Public

>     Functions and data types declared in header files located in :zephyr_file:`include/zephyr/`.

Reviewer: There was a proposal from @henrikbrixandersen or @nashif to use Doxygen as the ultimate source of truth instead of the location of the header. I personally prefer using the location, but I do agree that it is non-trivial to do so, because sometimes you need different API types to coexist in the same folder.
If we go with Doxygen classification, then this whole section should change to either remove paths altogether, or else at least highlight that the criteria is Doxygen-based and not path-based

   * - Prefix
     - Description
     - Example
   * - atomic_
     - Denotes an atomic operation.
     - `atomic_inc`
   * - device_
     - Denotes an API relating to devices and their initialization.
     - `device_get_binding`
   * - irq_
     - Denotes an IRQ management operation.
     - `irq_disable`
   * - isr_
     - Denotes an operation called by an Interrupt Service Routine.
     - `isr_event_send`
   * - k_
     - Kernel-specific function.
     - `k_malloc`
   * - k_do_
     - Kernel-specific functions indicating essential operation within the kernel space.
     - `k_do_event_signal`
   * - sys_
     - Catch-all for APIs that do not fit into the other namespaces.
     - `sys_write32`

Kconfig symbols
===============

> - Private - symbols that are promptless and hidden

Reviewer 2: There is an existing pattern that does not fit this definition is the *_PROMPTLESS set of Kconfigs. See:
- `$workspace/zephyr/modules/mbedtls/Kconfig` for MBEDTLS_PROMPTLESS;

used as :ref:`modules`may define their own rules for API lifecycle.

Zephyr is a constantly evolving project and API classification may change over time. A Private or
Internal API may be promoted to Internal or Public API, respectively. Zephyr users are encouraged to
follow :ref:`rfcs` process to recommend changes in API classification.

The following sections provide guidelines on how to identify the class of an API depending on its
type.

Header files contents
=====================
- Private - functions and data types declared in header files located in `include/zephyr/private`.
  In addition, private symbols are prefixed with `z_`. Due to historical reasons some APIs prefixed
  with `z_` are public.
- Internal - functions and data types declared `include/zephyr/internal`. In addition, Internal APIs
  must use @internal doxygen command.
- Public - functions and data types declared in header files located in `include/zephyr`.

In addition, the following prefixes are reserved by Zephyr kernel for use in Zephyr Public APIs:

.. list-table:: Prefixes and Descriptions
   :header-rows: 1
   :widths: 10 40 40
   :stub-columns: 1

   * - Prefix
     - Description
     - Example

>    * - `atomic_`

Reviewer 3: please check across the whole changeset, as there are other places where this comments applies I believe

.. _api_conventions.rst:

API Conventions
###############

Zephyr provides a few programming interfaces that users can use for creating their applications.
Zephyr's API space includes the following areas:
- C programming interface (function prototypes, structures, and macros), provided in C header files
- Configuration system (Kconfig)

> - Device driver model (Devicetree)

Reviewer 4: nitpicking here but I don't think it's fair to describe devicetree as "device driver model", that is more like the struct device and the whole story associated with it, the device tree is more like a hardware description language, should we find a better wording? https://en.wikipedia.org/wiki/Devicetree

Author: How about "Hardware description system (Devicetree)"

Reviewer 4: yeah sounds more like it

> Other areas, such as the build system (CMake), may also be seen as Zephyr's API. However, any area

Reviewer 5: How about zephyr specific CMake commands like `zephyr_library()` and `zephyr_library_sources_ifdef`? These are zephyr specific APIs right?

Author: Yes these are an API. However, we decided to limit the scope to items listed above. The motivation is to keep the scope narrow for now to make some progress and see if the policy works. We will improve over time.

that is not explicitly listed herein is not considered Zephyr's API as of today.

API classification
******************
The files that define the Zephyr APIs may contain symbols intended for different usage. The intended
API usage is designated by the API class. Zephyr APIs are classified as:

- Private
    These APIs are intended for use within boundary of a :term:`software component`. Private APIs
    defined in Zephyr tree are not subject to :ref:`api_lifecycle`. Therefore, can be changed or
    removed at any time. Changes to the private API may not be documented.

- Internal
    In general, these APIs are intended for use only between certain software components that are
    located in Zephyr tree. The context where the API is called or implemented is well defined. For
    example, functions prefixed with `arch_` are intended for use by the Zephyr kernel to invoke
    architecture specific code. An API is classified as an Internal on a case by case basis.

>     Internal APIs should not be used out-of-tree code. Internal APIs are not subject to

Reviewer 6: ```suggestion
    Internal APIs should not be used by out-of-tree code. Internal APIs are not subject to
```

>     Internal APIs should not be used out-of-tree code. Internal APIs are not subject to

Reviewer 4: may be worth dropping the specific example, I'd imagine a downstream architecture implementation may use/implement some `arch_` calls or something and it's odd to say that they shouldn't use it, there's not alternative. What it's fair to say is that that's the boundary where they should be a ware that they are building on an api that is not considered stable, with all the implications of that.

Author: As per process WG 2023-09-13 meeting:

@nashif - officially out of tree architectures are not supported. Internal APIs will not be subject to a lifecycle. There are many examples using internal APIs that need to be fixed.

    APIs must be documented in the migration guide.

- Public
    These APIs intended for use from any :term:`software component`. Public APIs may be used in-tree
    and out-of-tree. Public APIs are subject to the :ref:`api_lifecycle`. Therefore, changes to an
    API are introduced and documented according to the rules defined for the API life state.

Note, that only APIs defined in Zephyr tree are subject to :ref:`api_lifecycle`. External projects
used as :ref:`modules`may define their own rules for API lifecycle.

Zephyr is a constantly evolving project and API classification may change over time. A Private or
Internal API may be promoted to Internal or Public API, respectively. Zephyr users are encouraged to
follow :ref:`rfcs` process to recommend changes in API classification.

The following sections provide guidelines on how to identify the class of an API depending on its
type.

Header files contents
=====================
- Private - functions and data types declared in header files located in `include/zephyr/private`.
  In addition, private symbols are prefixed with `z_`. Due to historical reasons some APIs prefixed
  with `z_` are public.

> - Internal - functions and data types declared `include/zephyr/internal`. In addition, Internal APIs

Reviewer 4: sounds like a giant pain to manage, do we really need the two header namespaces?

Author: That was consensus from process WG meeting held on 20-09-2023. Please bring this topic to the process WG meeting if you feel strong about revising this.

  must use @internal doxygen command.
- Public - functions and data types declared in header files located in `include/zephyr`.

In addition, the following prefixes are reserved by Zephyr kernel for use in Zephyr Public APIs:

.. list-table:: Prefixes and Descriptions
   :header-rows: 1
   :widths: 10 40 40
   :stub-columns: 1

   * - Prefix
     - Description
     - Example
   * - atomic_
     - Denotes an atomic operation.
     - `atomic_inc`
   * - device_
     - Denotes an API relating to devices and their initialization.
     - `device_get_binding`
   * - irq_
     - Denotes an IRQ management operation.
     - `irq_disable`
   * - isr_
     - Denotes an operation called by an Interrupt Service Routine.
     - `isr_event_send`
   * - k_
     - Kernel-specific function.
     - `k_malloc`
   * - k_do_
     - Kernel-specific functions indicating essential operation within the kernel space.
     - `k_do_event_signal`
   * - sys_
     - Catch-all for APIs that do not fit into the other namespaces.
     - `sys_write32`

Kconfig symbols
===============

> Kconfig symbols that are promptless and hidden are Internal. All other symbols are Public. The

Reviewer 4: that's a really poor signal for the intent of the symbol, it usually just means that the symbol is a proxy for something but these are used for stuff like the i2c bit bang library, the keyboard scanning library, the M_CAN and SJA1000 support libraries... it would not be fair to just trash those symbols just because they have no symbols. This was talked about at length in the arch wg and there' really no easy way out at this point on such a flat namespace.

Author: Could you please be more specific what is your suggestion exactly?

Reviewer 4: Some hidden symbol should be considered stable as they are very much usable by the application. I don't really have a suggestion of what should not be considered stable though, the config system does not really help with that.

Reviewer 6: Should the language here just indicate that "Kconfig symbols are Public"?

Author: I have no objections to that.

Reviewer 6: Please make the edits here to indicate that all Kconfig symbols are Public.

Other areas, such as the build system (CMake), may also be seen as Zephyr's API. However, any area
that is not explicitly listed herein is not considered Zephyr's API as of today.

API classification
******************
The files that define the Zephyr APIs may contain symbols intended for different usage. The intended
API usage is designated by the API class. Zephyr APIs are classified as:

Private
    These APIs are intended for use within boundary of a :term:`software component`. Private APIs
    defined in Zephyr tree are not subject to :ref:`api_lifecycle`. Therefore, can be changed or
    removed at any time. Changes to the private API may not be documented.

Internal
    In general, these APIs are intended for use only between certain software components that are
    located in Zephyr tree. The context where the API is called or implemented is well defined. For
    example, functions prefixed with `arch_` are intended for use by the Zephyr kernel to invoke
    architecture specific code. An API is classified as an Internal on a case by case basis.
    Internal APIs should not be used by out-of-tree code. Internal APIs are not subject to

>     ref:`api_lifecycle`. Therefore, can be changed or removed at any time. Changes to the internal

Reviewer 6: Broken link
```suggestion
   :ref:`api_lifecycle`. Therefore, can be changed or removed at any time. Changes to the internal
```

    APIs must be documented in the migration guide.

Public
    These APIs intended for use from any :term:`software component`. Public APIs may be used in-tree
    and out-of-tree. Public APIs are subject to the :ref:`api_lifecycle`. Therefore, changes to an
    API are introduced and documented according to the rules defined for the API life state.

Note, that only APIs defined in Zephyr tree are subject to :ref:`api_lifecycle`. External projects
used as :ref:`modules` may define their own rules for API lifecycle.

Zephyr is a constantly evolving project and API classification may change over time. A Private or
Internal API may be promoted to Internal or Public API, respectively. Zephyr users are encouraged to
follow :ref:`rfcs` process to recommend changes in API classification.

The following sections provide guidelines on how to identify the class of an API depending on its
type.

Header files contents
=====================
Private

>     Functions and data types declared in header files located in `include/zephyr/private`. In

Reviewer 6: ```suggestion
    Functions and data types declared in header files located in :zephyr_file:`include/zephyr/private/`. In
```

>     Functions and data types declared in header files located in
>     :zephyr_file:`include/zephyr/private/`. In addition, private symbols are prefixed with ``z_``.
>     Due to historical reasons some APIs prefixed with ``z_`` are public.

Reviewer 6: The path `include/zephyr/private` doesn't exist, so the link doesn't work.  Maybe this should read 
```suggestion
    Functions and data types declared in header files located outside of 
    :zephyr_file:`include/zephyr/` and :zephyr_file`include/zephyr/internal` are private. 
    In addition, private symbols are prefixed with ``z_``.
    Due to historical reasons some APIs prefixed with ``z_`` are public.
```

Author: Perhaps I should just create `include/zephyr/private` instead?

Reviewer 6: Fundamentally, if an API is private it doesn't belong under `include/zephyr`.    There are probably a few exceptions, but hopefully the `z_` prefix covers those cases.

I'm in favor of designating anything outside of `include/zephyr` as private.

Internal
    Functions and data types declared `include/zephyr/internal`. In addition, Internal APIs must

>     use @internal doxygen command.

Reviewer 6: ```suggestion
    use ``@internal`` doxygen command.
```

>     Functions and data types declared `include/zephyr/internal`. In addition, Internal APIs must
>     use ``@internal`` doxygen command.

Reviewer 6: ```suggestion
    Functions and data types declared :zephyr_file:`include/zephyr/internal`. In addition, Internal APIs must
    use ``@internal`` doxygen command.
```

Public

>     Functions and data types declared in header files located in `include/zephyr`.

Reviewer 6: ```suggestion
    Functions and data types declared in header files located in :zephyr_file:`include/zephyr/`.
```

In addition, the following prefixes are reserved by Zephyr kernel for use in Zephyr Public APIs:

.. list-table:: Prefixes and Descriptions
   :header-rows: 1
   :widths: 10 40 40
   :stub-columns: 1

   * - Prefix
     - Description
     - Example
   * - ``atomic_``
     - Denotes an atomic operation.
     - ``atomic_inc``
   * - ``device_``
     - Denotes an API relating to devices and their initialization.

>      - ``device_get_binding``

Reviewer 6: Please add reference for all the functions in this table.  Looks like atomic_inc is missing the doxygen tags and may not work.
```suggestion
     - :c:func:`device_get_binding`
```

   * - ``irq_``
     - Denotes an IRQ management operation.

>      - ``irq_disable``

Reviewer 6: Please add `:c:func:` references for all the routines that have valid doxygen blocks.

Author: In addition I have removed these prefixes where I didn't find any function that would use them.

.. _api_conventions.rst:

API Conventions
###############

Zephyr provides a few programming interfaces that users can use for creating their applications.
Zephyr's API space includes the following areas:
- C programming interface (function prototypes, structures, and macros), provided in C header files
- Configuration system (Kconfig)
- Device driver model (Devicetree)

Other areas, such as the build system (CMake), may also be seen as Zephyr's API. However, any area
that is not explicitly listed herein is not considered Zephyr's API as of today.

API classification
******************
The files that define the Zephyr APIs may contain symbols intended for different usage. The intended
API usage is designated by the API class. Zephyr APIs are classified as:

> - private - these APIs are intended for use in-tree within boundary of a software component such as

Reviewer 2: API usage and API class is not limited to just within the Zephyr repository.  Please:

- Remove "in-tree" from the definition so that it is a broadly applicable definition;
- Qualify the lifecycle statement so that it applies to only those APIs defined in the Zephyr repository.

Author: Done.

            driver or subsystem. Private APIs are not subject to :ref:`api_lifecycle`. Therefore can
            be changed or removed at any time.

> - internal - a class of APIs that serve as the interface between two or more low level areas.

Reviewer 2: Please enumerate which areas are meant by "low level". It would probably be helpful if Zephyr Project first documented which Zephyr directory tree subdirectories are not intended to be extended in other modules (e.g., arch; kernel; others?). Once listed then the interfaces between these subtrees would already be internal to the Zephyr repository.

Author: I have removed "low level". Please see what you think about the new defintion.

> - public - a class of APis intended for use in-tree and out-of-tree. The public APIs are subject to

Reviewer 2: Please qualify the api lifecycle statement to those APIs defined in the Zephyr repository unless we are also going to apply it to APIs of depended-upon modules (e.g., TFM, mbedtls, mcuboot).

Author: Added a sentence to clarify.

           the :ref:`api_lifecycle`. This is the class that most users should use.

The following sections provide guidelines on how to identify the class of an API depending on its
type.

Header files contents
=====================
- private - symbols prefixed with `z_` are private.
- internal - <TBD>
- public - functions and data types declared in header files located in `include/zephyr`.

Due to historical reasons some APIs prefixed with `z_` are public. These are:
- <TBD>

Kconfig symbols
===============
- private - not applicable

> - internal - symbols that are promptless and hidden

Reviewer 2: 'internal' here should have the same scope as 'internal' does for APIs.  I think this means that symbols that are promptless and hidden are "private" rather than "internal".

Author: Agree.

.. _api_conventions.rst:

API Conventions
###############

Zephyr provides a few programming interfaces that users can use for creating their applications.
Zephyr's API space includes the following areas:
- C programming interface (function prototypes, structures, and macros), provided in C header files
- Configuration system (Kconfig)
- Device driver model (Devicetree)

Other areas, such as the build system (CMake), may also be seen as Zephyr's API. However, any area
that is not explicitly listed herein is not considered Zephyr's API as of today.

API classification
******************
The files that define the Zephyr APIs may contain symbols intended for different usage. The intended
API usage is designated by the API class. Zephyr APIs are classified as:

- private - these APIs are intended for use within boundary of a software component such as

>             driver or subsystem. Private APIs defined in Zephyr tree are not subject to

Reviewer 2: I think we need an actual definition for "software component". AFAICT we intend this to mean the smallest unit we control to compose a system. It seems pretty clear that each shell backend is a different component. Since the scheduler is removed from the kernel in a minimal configuration it might also be one or more components.

I am hesitant to equate component with unit because the "unit_testing" board uses the host toolchain rather than the target toolchain, and it is very reasonable to have a separate unit test suite for each file in a component.

> - private - these APIs are intended for use within boundary of a software component such as
>             driver or subsystem. Private APIs defined in Zephyr tree are not subject to
>             :ref:`api_lifecycle`. Therefore, can be changed or removed at any time.

Reviewer 3: The indenting is wrong, it should be either the following to render as a "description list"

```suggestion
- Private
          These APIs are intended for use within boundary of a software component such as
          driver or subsystem. Private APIs defined in Zephyr tree are not subject to
          :ref:`api_lifecycle`. Therefore, can be changed or removed at any time.
```

Or, to render as a regular bullet point, everything aligned after the bullet and whitespace.

```suggestion
- Private - These APIs are intended for use within boundary of a software component such as
   driver or subsystem. Private APIs defined in Zephyr tree are not subject to
   :ref:`api_lifecycle`. Therefore, can be changed or removed at any time.
```

BTW you can see the rendered doc here:  https://builds.zephyrproject.io/zephyr/pr/62198/docs/develop/api/api_conventions.html#api-conventions

Author: Done

- internal - these APIs are intended for use only between certain software components. The context 
            where the API is called or implemented is well defined. For example, functions prefixed
            with `arch_` are intended for use by the Zephyr kernel to invoke

>             architecture specific code. Internal APIs may be used in-tree and out-of-tree. Internal

Author: Done.

>             APIs defined in Zephyr tree are not subject to :ref:`api_lifecycle`. Therefore, can be changed or removed at any time.

Reviewer 3: The statement "Therefore, can be changed or removed at any time." is the exact same as for private, but I think this one needs more boundaries, maybe? -- yes it may be changed or removed at any time, but it should still be documented/tracked somewhere (release notes / migration plan?) as there is a much higher chance that such a change would break "internal" users.

Author: Added a note about migration guide.

- public - these APIs intended for use from any software component. Public APIs may be used in-tree
            and out-of-tree. Public APIs are subject to the :ref:`api_lifecycle`. Therefore, 
            changes to an API are done according to the rules defined for the API life state.

> Note, that only APIs defined in Zephyr tree are subject to :ref:`api_lifecycle`. Independant 
> projects, such as TF-M or MCUboot, may define their own rules for API lifecycle.

Reviewer 6: ```suggestion
Note, that only APIs defined in Zephyr tree are subject to :ref:`api_lifecycle`. External modules (TBD link) may define their own rules for an API lifecycle.
```

Author: Done.

.. _api_conventions.rst:

API Conventions
###############

Zephyr provides a few programming interfaces that users can use for creating their applications.
Zephyr's API space includes the following areas: - C programming interface (function prototypes,
structures, and macros), provided in C header files - Configuration system (Kconfig) - Device driver
model (Devicetree)

Other areas, such as the build system (CMake), may also be seen as Zephyr's API. However, any area
that is not explicitly listed herein is not considered Zephyr's API as of today.

API classification
******************
The files that define the Zephyr APIs may contain symbols intended for different usage. The intended
API usage is designated by the API class. Zephyr APIs are classified as:

- Private
    These APIs are intended for use within boundary of a :term:`software component`. Private APIs
    defined in Zephyr tree are not subject to :ref:`api_lifecycle`. Therefore, can be changed or
    removed at any time. Changes to the private API may not be documented.

> - Internal

Reviewer 6: From Process WG: Suggest adding language indicating that selection of APIs into public/internal classes is done on a case by case basis.

Reviewer 6: Also add a sentence indicated that internal APIs can get promoted to public if there is justification (user creates an RFC/PR to propose the promotion.

    These APIs are intended for use only between certain software components. The context

>     where the API is called or implemented is well defined. For example, functions prefixed with

Reviewer 6: I just came across Zephyr naming convention wiki: https://github.com/zephyrproject-rtos/zephyr/wiki/Naming-Conventions

That doc likely needs updating for consistency with this doc.

Author: I'd suggest removing (or move to docs) this wiki page altogether.
